### ISSUE TITLE
Add a priority that means "block until this lands"

### ISSUE THREAD

**Issue by larsbergstrom** [servo/homu#111]
There are several scenarios in which we'd like to be able to abort / stop all other patches from landing until a given PR has made it through. Including:
- Easily bitrotted but high-value PRs (e.g., updates to lots of key Rust components)
- Massive refactorings
- A backout of change that is tree-breaking

I'd propose something like `r+ p=tree-closing`. There are at least two parts to this:
1) The internal mechanics of ensuring that no other PR lands (and anything in-progress is aborted, a lá `force`) while something with this priority is in the queue
2) Ensuring the queue display page clearly reports the reason that the queue is not processing, especially when something like a build failure happens in that PR but the queue is still blocked

cc @edunham @Manishearth @globau

(I'd especially like bikeshedding on the name of the priority, as I'm sure there's something pithy but I'm missing it)

**Comment by Manishearth** [comment]
This basically exists already.

`@homu treeclosed=5` will close the tree for any PR below priority 5. (It can be unset with `treeclosed-`)

It will not however interrupt existing builds or prevent them from landing. This support could be added I guess?

It does show tree closed statuses on the queue.

I've used to a few times, works great.

**Comment by Manishearth** [comment]
See https://github.com/servo/homu/pull/87

**Comment by larsbergstrom** [comment]
@Manishearth Yeah, the additional features (`force` and really ignoring any other commands from other PRs) are the key bits required for the "mozilla-central backout with corresponding PR against servo/servo" scenario.

**Comment by globau** [comment]
would it be possible to restrict usage of highest priority?  a backout shouldn't be able to be trumped by a user pushing a higher priority.

**Comment by edunham** [comment]
The above 2 PRs implement a cap on possible priority values. 

@globau, do you mean that there should be a limited list of users permitted to use the top priority? 

As things stand immediately after that pair of changes lands, a sheriff can ensure that their PR goes through before anyone else by closing the tree to builds below p=9000, setting their change to p=9001, and walking away. If a subsequent change came in which was also p=9001, it would have to wait until the first p=9001 PR finished. If the second change was meant to supersede the first max-priority build, someone would have to tell Homu to de-prioritize or cancel the first one.

**Comment by globau** [comment]
@edunham nice - setting a cap on priority values is an elegant solution to this issue.  i agree that all p=9001 jobs should be processed as a queue.

**Comment by edunham** [comment]
From chatting with @bholley about tree closure and backouts on IRC: 

A backout is a revert commit. 

If the sheriffs want to back out a bunch of stuff at once, they can close the tree with `treeclosed=9000`. Then they set all jobs which need to run while the tree is closed as `p=9001`. It is at their discretion whether they run all backouts as a single PR, or do a rollup, or enqueue several PRs in a particular order (PRs enter the queue in the order that they get r+'d with the given priority) and let them run sequentially. 

I hope that someday in the future we can get fancier rollup logic similar to what Gerrit uses, but that's months to years out. For the time being, a single PR and a rollup will be roughly the same speed, whereas a series of sequentially running PRs will be slower to complete but working changesets will land faster if a single changeset in the group fails.

**Comment by Eh2406** [comment]
move to close?
